File: apps/api/src/platform/db/pool.ts
import pg from 'pg';

/**
 * The PostgreSQL connection pool.
 *
 * Client management is the first module that genuinely persists: a customer
 * record has to survive a restart, which the in-memory MES services do not.
 * The pool is created lazily so an install without `DATABASE_URL` still boots,
 * with the client-management endpoints reporting that they are unavailable
 * rather than the whole API refusing to start.
 */

let pool: pg.Pool | null = null;
let warned = false;
let rlsBypassed = false;

export function databaseUrl(): string | undefined {
  return process.env.DATABASE_URL || undefined;
}

export function isDatabaseConfigured(): boolean {
  return Boolean(databaseUrl());
}

export function getPool(): pg.Pool {
  const url = databaseUrl();
  if (!url) {
    throw new Error('DATABASE_URL is not set, so no database connection is available.');
  }
  if (!pool) {
    pool = new pg.Pool({
      connectionString: url,
      // A small pool: this API is one process serving one plant, and a large
      // pool would take connections a shared Postgres needs for other work.
      max: Number(process.env.DATABASE_POOL_MAX ?? 10),
      idleTimeoutMillis: 30_000,
      connectionTimeoutMillis: 10_000,
    });
    pool.on('error', (error) => {
      // A pooled connection can die while idle; log it rather than letting an
      // unhandled 'error' event take the process down.
      // eslint-disable-next-line no-console
      console.error('[db] idle client error:', error.message);
    });
  }
  return pool;
}

/** Runs a query, returning the rows. */
export async function query<T extends pg.QueryResultRow = pg.QueryResultRow>(
  text: string,
  params: unknown[] = []
): Promise<T[]> {
  const result = await getPool().query<T>(text, params);
  return result.rows;
}

/** Runs a query expecting at most one row. */
export async function queryOne<T extends pg.QueryResultRow = pg.QueryResultRow>(
  text: string,
  params: unknown[] = []
): Promise<T | undefined> {
  const rows = await query<T>(text, params);
  return rows[0];
}

/**
 * Runs several statements in one transaction.
 *
 * Creating a client touches `tenant`, `client_account` and
 * `client_subscription`; a half-created customer is worse than a failed
 * request, so those go in together or not at all.
 */
export async function transaction<T>(fn: (client: pg.PoolClient) => Promise<T>): Promise<T> {
  const client = await getPool().connect();
  try {
    await client.query('BEGIN');
    const result = await fn(client);
    await client.query('COMMIT');
    return result;
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
}

/**
 * Runs `fn` with the tenant declared to PostgreSQL, so the `tenant_isolation`
 * policies on every tenant-scoped table apply to its queries (§13).
 *
 * The setting has to be transaction-local, hence the wrapping transaction:
 * `SET LOCAL` outside one either errors or leaks the tenant onto the next
 * borrower of that pooled connection, which is the worst possible failure mode
 * for an isolation control. `set_config` is used rather than `SET LOCAL
 * app.tenant_id = '...'` because the tenant id can then be a bind parameter;
 * `SET` takes no parameters and would mean interpolating it into SQL.
 *
 * This is defence in depth, not a replacement for scoping queries in the
 * application: it is what stops a query that forgot its `WHERE tenant_id`
 * from returning another factory's production.
 */
export async function withTenant<T>(
  tenantId: string,
  fn: (client: pg.PoolClient) => Promise<T>
): Promise<T> {
  if (!tenantId) throw new Error('withTenant requires a tenant id.');
  return transaction(async (client) => {
    await client.query('SELECT set_config($1, $2, true)', ['app.tenant_id', tenantId]);
    return fn(client);
  });
}

/** Verifies the database is reachable and the client tables are present. */
export async function checkDatabase(): Promise<{ ok: boolean; detail: string }> {
  if (!isDatabaseConfigured()) {
    if (!warned) {
      warned = true;
      // eslint-disable-next-line no-console
      console.warn('[db] DATABASE_URL is not set. Client management is unavailable.');
    }
    return { ok: false, detail: 'DATABASE_URL is not set' };
  }
  try {
    const row = await queryOne<{ count: string }>(
      "SELECT count(*)::text AS count FROM information_schema.tables WHERE table_name = 'client_account'"
    );
    if (row?.count === '0') {
      return { ok: false, detail: 'client_account table is missing, run pnpm db:migrate' };
    }

    // Tenant isolation is only real if the connecting role is subject to the
    // policies. A superuser (or any role with BYPASSRLS) is exempt from all of
    // them, so the tables would look protected while every tenant could read
    // every other tenant's rows. Checked at startup because the failure is
    // invisible in normal operation, right up to the moment it is a breach.
    const role = await queryOne<{ superuser: boolean; bypassrls: boolean; name: string }>(
      'SELECT rolname AS name, rolsuper AS superuser, rolbypassrls AS bypassrls FROM pg_roles WHERE rolname = current_user'
    );
    if (role && (role.superuser || role.bypassrls)) {
      rlsBypassed = true;
      // eslint-disable-next-line no-console
      console.error(
        `[db] SECURITY: connected as "${role.name}", which bypasses row-level security ` +
          `(superuser=${role.superuser}, bypassrls=${role.bypassrls}). Tenant isolation policies ` +
          'do NOT apply to this connection. Run pnpm db:migrate with APP_DB_PASSWORD set and point ' +
          'DATABASE_URL at the factory_app role.'
      );
      return { ok: true, detail: `connected as ${role.name}, ROW LEVEL SECURITY BYPASSED` };
    }

    return { ok: true, detail: 'connected' };
  } catch (error) {
    return { ok: false, detail: error instanceof Error ? error.message : 'connection failed' };
  }
}

/** Whether the connecting role is exempt from the tenant policies. */
export function isRlsBypassed(): boolean {
  return rlsBypassed;
}

export async function closePool(): Promise<void> {
  if (pool) {
    await pool.end();
    pool = null;
  }
}

File: apps/api/internal/platform/db/pool.go
// Package db owns the PostgreSQL connection pool and the two ways a request is
// allowed to touch it: inside a tenant-scoped transaction, or - for the relay
// and the job runner only - without a tenant at all.
package db

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// Options tunes the pool. Zero values fall back to what a single API process
// serving one plant needs; a large pool would take connections a shared
// PostgreSQL needs for other work.
type Options struct {
	MaxConns          int32
	MinConns          int32
	MaxConnLifetime   time.Duration
	MaxConnIdleTime   time.Duration
	HealthCheckPeriod time.Duration
	// SlowQuery, when > 0, logs any statement that takes longer.
	SlowQuery time.Duration
	// Observer receives every statement's duration; nil means metrics off.
	Observer QueryObserver
	Logger   *slog.Logger
}

// Pool wraps pgxpool with the tenant discipline every caller must follow.
type Pool struct {
	*pgxpool.Pool
	log *slog.Logger
}

// Open connects and verifies the connection.
func Open(ctx context.Context, url string, opts Options) (*Pool, error) {
	cfg, err := pgxpool.ParseConfig(url)
	if err != nil {
		return nil, fmt.Errorf("DATABASE_URL: %w", err)
	}
	if opts.MaxConns > 0 {
		cfg.MaxConns = opts.MaxConns
	}
	if opts.MinConns > 0 {
		cfg.MinConns = opts.MinConns
	}
	cfg.MaxConnLifetime = orDuration(opts.MaxConnLifetime, 30*time.Minute)
	cfg.MaxConnIdleTime = orDuration(opts.MaxConnIdleTime, 5*time.Minute)
	cfg.HealthCheckPeriod = orDuration(opts.HealthCheckPeriod, time.Minute)
	cfg.ConnConfig.ConnectTimeout = 10 * time.Second
	// Prepared statements are cached per connection by pgx itself; a
	// statement the API runs on every request is parsed by PostgreSQL once
	// per connection rather than once per call.
	cfg.ConnConfig.DefaultQueryExecMode = pgx.QueryExecModeCacheStatement

	log := opts.Logger
	if log == nil {
		log = slog.Default()
	}
	cfg.ConnConfig.Tracer = &tracer{slow: opts.SlowQuery, observer: opts.Observer, log: log}

	pool, err := pgxpool.NewWithConfig(ctx, cfg)
	if err != nil {
		return nil, err
	}
	if err := pool.Ping(ctx); err != nil {
		pool.Close()
		return nil, fmt.Errorf("database unreachable: %w", err)
	}
	return &Pool{Pool: pool, log: log}, nil
}

// RoleStatus is what CheckRole learns about the connecting role.
type RoleStatus struct {
	Name      string
	Superuser bool
	BypassRLS bool
}

// Bypassed is true when tenant isolation policies do not apply to this role.
func (s RoleStatus) Bypassed() bool { return s.Superuser || s.BypassRLS }

// CheckRole reads the connecting role's attributes. Tenant isolation is only
// real if the role is subject to the policies; a superuser is exempt from all
// of them, so the tables would look protected while every tenant could read
// every other tenant's rows. Checked at startup because the failure is
// invisible in normal operation, right up to the moment it is a breach.
func (p *Pool) CheckRole(ctx context.Context) (RoleStatus, error) {
	var s RoleStatus
	err := p.QueryRow(ctx,
		`SELECT rolname, rolsuper, rolbypassrls FROM pg_roles WHERE rolname = current_user`,
	).Scan(&s.Name, &s.Superuser, &s.BypassRLS)
	return s, err
}

// TableExists answers whether the schema has been migrated far enough to
// hold a given table.
func (p *Pool) TableExists(ctx context.Context, table string) (bool, error) {
	var n int
	err := p.QueryRow(ctx,
		`SELECT count(*) FROM information_schema.tables WHERE table_name = $1`, table,
	).Scan(&n)
	return n > 0, err
}

// ErrTenantRequired is returned when a caller forgets the tenant: a query
// that runs without one either sees nothing (RLS) or, worse, would run
// with whatever the previous borrower of the connection declared.
var ErrTenantRequired = errors.New("db: WithTenant requires a tenant id")

// WithTenant runs fn inside a transaction with the tenant declared to
// PostgreSQL, so the tenant_isolation policies on every tenant-scoped table
// apply to its queries.
//
// The setting has to be transaction-local, hence the transaction: SET LOCAL
// outside one either errors or leaks the tenant onto the next borrower of that
// pooled connection, which is the worst possible failure mode for an isolation
// control. set_config is used so the tenant id can be a bind parameter.
//
// This is defence in depth, not a replacement for scoping queries in the
// application: it is what stops a query that forgot its WHERE tenant_id
// from returning another factory's production.
func (p *Pool) WithTenant(ctx context.Context, tenantID string, fn func(tx pgx.Tx) error) error {
	if tenantID == "" {
		return ErrTenantRequired
	}
	return p.transaction(ctx, func(tx pgx.Tx) error {
		if _, err := tx.Exec(ctx, `SELECT set_config('app.tenant_id', $1, true)`, tenantID); err != nil {
			return err
		}
		return fn(tx)
	})
}

// WithoutTenant runs fn in a transaction with no tenant declared. Only the
// outbox relay and the job queue may use it: planning_job and outbox_event
// carry a runner_without_tenant policy (migration 034) that admits exactly
// this case, and every other table shows such a transaction nothing.
func (p *Pool) WithoutTenant(ctx context.Context, fn func(tx pgx.Tx) error) error {
	return p.transaction(ctx, fn)
}

func (p *Pool) transaction(ctx context.Context, fn func(tx pgx.Tx) error) (err error) {
	tx, err := p.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() {
		if r := recover(); r != nil {
			_ = tx.Rollback(ctx)
			panic(r)
		}
	}()
	if err := fn(tx); err != nil {
		if rbErr := tx.Rollback(ctx); rbErr != nil && !errors.Is(rbErr, pgx.ErrTxClosed) {
			p.log.Warn("db: rollback failed", "error", rbErr)
		}
		return err
	}
	return tx.Commit(ctx)
}

// Notify raises pg_notify inside the caller's transaction, so listeners
// only hear about a change once it has committed.
func Notify(ctx context.Context, tx pgx.Tx, channel, payload string) error {
	_, err := tx.Exec(ctx, `SELECT pg_notify($1, $2)`, channel, payload)
	return err
}

func orDuration(v, fallback time.Duration) time.Duration {
	if v > 0 {
		return v
	}
	return fallback
}
